mindmeld/text_preparation/text_preparation_pipeline.py

```diff
@@ -0,0 +1,530 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2015 Cisco Systems, Inc. and others.  All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This module contains a text Processing Pipeline."""
+import logging
+from typing import List, Dict
+import re
+import unicodedata
+
+from .preprocessors import Preprocessor, PreprocessorFactory
+from .normalizers import (
+    Normalizer,
+    NoOpNormalizer,
+    NormalizerFactory,
+    RegexNormalizer,
+    Lowercase,
+    ASCIIFold,
+)
+from .tokenizers import Tokenizer, TokenizerFactory, WhiteSpaceTokenizer
+from .stemmers import Stemmer, StemmerFactory, EnglishNLTKStemmer
+
+from ..components._config import (
+    get_text_preparation_config,
+    get_language_config,
+    ENGLISH_LANGUAGE_CODE,
+)
+from ..constants import UNICODE_SPACE_CATEGORY, DEFAULT_REGEX_NORM_RULES
+
+logger = logging.getLogger(__name__)
+
+
+# Regex Pattern to capture MindMeld entities ("{entity_text|entity_type|optional_role}")
+MINDMELD_ANNOTATION_PATTERN = r"\{([^\}\|]*)\|[^\{]*\}"
+
+
+class TextPreparationPipelineError(Exception):
+    pass
+
+
+class TextPreparationPipeline:
+    """Pipeline Class for MindMeld's text processing."""
+
+    def __init__(
+        self,
+        tokenizer: Tokenizer,
+        stemmer: Stemmer,
+        preprocessors: List[Preprocessor] = None,
+        normalizers: List[Normalizer] = None,
+        language: str = ENGLISH_LANGUAGE_CODE,
+    ):
+        """Creates a Pipeline instance."""
+        self.language = language
+        self.preprocessors = preprocessors
+        self.normalizers = normalizers or [NoOpNormalizer()]
+        self.tokenizer = tokenizer
+        self.stemmer = stemmer
```

Comment:
Do we also want to use a `noopstemmer` here as an or condition? Same with the other noop components

Reply:
Sure, added NoOp defaults for the stemmer and preprocessors:

```
self.preprocessors = preprocessors or [NoOpPreprocessor()]
self.normalizers = normalizers or [NoOpNormalizer()]
self.tokenizer = tokenizer
self.stemmer = stemmer or NoOpStemmer()
```